src/StatusesHandler.php

```diff
@@ -0,0 +1,270 @@
+    use NormalizeArrayCapableTrait;
+
+    /**
+     * List of statuses keys available in application.
+     *
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $statuses;
+
+    /**
+     * List of statuses keys that are not visible in the UI.
+     *
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $hiddenStatuses;
+
+    /**
+     * Map of known statuses keys translations.
+     *
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $statusesLabels;
+
+    /**
+     * Option key name to save screen statuses config.
+     *
+     * @since [*next-version*]
+     *
+     * @var string
+     */
+    protected $screenOptionsKey;
+
+    /**
+     * Endpoint for saving statuses.
+     *
+     * @since [*next-version*]
+     *
+     * @var string
+     */
+    protected $statusesEndpoint;
+
+    /**
+     * StatusesHandler constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param Traversable           $statuses         List of statuses key in application.
+     * @param Traversable           $hiddenStatuses   List of hidden statuses in application.
+     * @param Traversable           $statusesLabels   Map of known statuses labels.
+     * @param string                $screenOptionsKey Option key name to save screen statuses config.
+     * @param string                $statusesEndpoint Endpoint for saving statuses.
+     * @param EventManagerInterface $eventManager     The event manager.
+     * @param EventFactoryInterface $eventFactory     The event factory.
+     */
+    public function __construct(
+        Traversable $statuses,
+        Traversable $hiddenStatuses,
+        Traversable $statusesLabels,
+        $screenOptionsKey,
+        $statusesEndpoint,
+        $eventManager,
+        $eventFactory
+    ) {
+        $this->statuses       = $this->_normalizeArray($statuses);
+        $this->hiddenStatuses = $this->_normalizeArray($hiddenStatuses);
+        $this->statusesLabels = $this->_normalizeArray($statusesLabels);
+
+        $this->screenOptionsKey = $screenOptionsKey;
+        $this->statusesEndpoint = $statusesEndpoint;
+
+        $this->_setEventManager($eventManager);
+        $this->_setEventFactory($eventFactory);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function __invoke()
+    {
+        /* @var $event EventInterface */
+        $event = func_get_arg(0);
+
+        if (!($event instanceof EventInterface)) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument is not an event instance'), null, null, $event
+            );
+        }
+
+        $event->setParams(array_merge(
+            $event->getParams(),
+            $this->_getStatusesParams()
+        ));
+    }
+
+    /**
+     * Attach events for status handling.
+     *
+     * @since [*next-version*]
+     */
+    public function run()
```

Comment:
What interface does this implement?

Reply:
Yes, I don't found interface for it. It would be good to discuss it in more details.

Reply:
As discussed, a good way to avoid this is by creating another handler, and attaching it in the module, like all the other handlers.